### PR TITLE
context: Add context support with ContextWithLogger and FromContext

### PR DESCRIPTION
**Pull request by Prnyself**

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+type contextKey struct{}
+
+// loggerKey is used as key to store logger in context
+var loggerKey contextKey
+
+// ContextWithLogger set *Logger into given context and return
+func ContextWithLogger(ctx context.Context, l *Logger) context.Context {
+	if ctx == nil {
```
Comment (Xuanwo) [inline context.go:14]: if `ctx == nil && l == nil` return `Background` directly.
if `l == nil` return ctx
if ctx already has a logger, return ctx directly

File: context.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		ctx = context.Background()
+	}
+	return context.WithValue(ctx, loggerKey, l)
+}
+
+// FromContext get *Logger from context
+// Notice: If ctx is nil or no Logger was set before, it will return nil
+func FromContext(ctx context.Context) *Logger {
+	if ctx == nil {
+		return nil
```
Comment (Xuanwo) [inline context.go]: if ctx is nil, return a basic logger
if ctx doesn't has a logger, return a basic logger

**Comment by Xuanwo** [comment]
This log is designed for critical mission so we neither return `nil` nor `panic` user's application, instead, we always return useable value for user.

File: context.go (hunk 1 above)
Comment (Prnyself) [reply context.go:14]: > if ctx already has a logger, return ctx directly

Should new logger replace the older one?

Comment (Xuanwo) [reply context.go:14]: Oh, yes, so we don't need to check this.

**Pull request merged**